### DISCUSSION TITLE
Frigate Tip: Motion Masks

### DISCUSSION THREAD

**Discussion by hawkeye217** [blakeblackshear/frigate#13040] [Tips]
The *Frigate Tips* series is a collection of posts where the Frigate developers give tips on how to best configure, use, and maintain your Frigate installation. This post's tip aims to bring clarity on when to use *motion masks*, something we've discovered is often misunderstood by new users and users coming from other NVR software.

### What is a motion mask?

*Motion masks* in Frigate are used to prevent areas of a scene on a camera from triggering motion detection. They are most easily configured from the Frigate Web UI. The Frigate [documentation](https://docs.frigate.video/configuration/masks) says:

>  You want to mask out your timestamp, the sky, rooftops, etc. Keep in mind that this mask only prevents motion from being detected and does not prevent objects from being detected if object detection was started due to motion in unmasked areas.

This is a really important point that we've found that many users misunderstand: *a motion mask only prevents motion from being detected and does not prevent objects from being detected*.

### Motion masks don't prevent objects from being detected

Motion masks should **not** be used to block out areas where you aren't interested in detecting things.

Yes, you read that correctly (though there are some caveats). Motion masks should be used *very sparingly*. In order to understand why, it's important to understand how Frigate works.

Again, from the Frigate [documentation](https://docs.frigate.video/configuration/masks#further-clarification):

> Frigate uses motion detection as a first line check to see if there is anything happening in the frame worth checking with object detection.

> Once motion is detected, it tries to group up nearby areas of motion together in hopes of identifying a rectangle in the image that will capture the area worth inspecting. These are the red "motion boxes" you see in the debug viewer.

For a simple example, take this scene of a street in Orlando, Florida. A car is driving along the road, and Frigate's Debug view shows red boxes to indicate that motion is being detected on the frame in the areas where the car is moving. These regions of motion are grouped together and passed onto Frigate's object detector, which identifies the object in the frame as a car. Frigate's Debug view then shows the label and colored bounding box around it.

![Debug view with bounding box](https://github.com/user-attachments/assets/f3ad704f-5f9e-448d-a709-5bb9eb0c221c)

Let's imagine a scenario with this scene where you're only interested in people that are walking on the sidewalk on the right side of the frame, and you only want those people to be shown in the Review pane in Frigate. 

It's often assumed that this is where a *motion mask* would be the solution. A new user may be tempted to add a motion mask like this:

![Invalid motion mask](https://github.com/user-attachments/assets/8947d264-c7a7-4fd4-abd0-a7b661ad691b)

However, this is **not** the case, and this kind of a motion mask is actually detrimental to the way Frigate is designed to work.

Again, from the [documentation](https://docs.frigate.video/configuration/masks#further-clarification):

> With too much masking, tracking is hampered and if an object walks from an unmasked area into a fully masked area, they essentially disappear and will be picked up as a "new" object if they leave the masked area. This is important because Frigate uses the history of scores while tracking an object to determine if it is a false positive or not.

So take this hypothetical situation:

A person walks underneath the clock (which, if using the masking example above, is in a masked area) and proceeds down the sidewalk we're interested in (which is now in the unmasked area).

![Person walking](https://github.com/user-attachments/assets/fd7959df-b13b-4cc9-9882-54cbdeafe79a)

Frigate would detect motion just outside of the masked area and begin tracking the person as they proceed down the sidewalk. 

Seems like this would work... But wait.

What if the person realized they had dropped something as they were walking underneath the clock? They would turn around and proceed *back into the masked area*. Perhaps they spend a couple of minutes there talking to someone else, and then continue back down the sidewalk of interest. 

Because of overmasking, the person of interest would actually be tracked by Frigate as an entirely new person as they leave the masked area and proceed down the sidewalk the second time, potentially creating two separate review items. In a high-traffic area, this would certainly not be ideal. 

So, what's the solution?

### Use `required_zones` instead of motion masks

If you're only interested in keeping track of objects in a particular area of your scene, use `required_zones`.

From the [documentation](https://docs.frigate.video/configuration/zones):

> Zones allow you to define a specific area of the frame and apply additional filters for object types so you can determine whether or not an object is within a particular area. Presence in a zone is evaluated based on the bottom center of the bounding box for the object. It does not matter how much of the bounding box overlaps with the zone.

Using a zone for the sidewalk and then restricting alerts and/or detections to objects in that zone is the correct solution here. This can easily be done with the Frigate UI (0.14 and later).

Create a zone called "sidewalk" in Frigate's mask/zone editor and then restart Frigate:

![Create a zone](https://github.com/user-attachments/assets/12e38033-c587-47d2-a136-e8c9be4d371d)

After restarting, navigate to Settings &rarr; Camera Settings, choose the camera in the top right, and then check the boxes to restrict alerts and/or detections to the "sidewalk" zone you just created. Then save and restart Frigate once more.

![Camera configuration](https://github.com/user-attachments/assets/c0bada4e-4bc6-424c-94b7-300a1d311232)

With this configuration, Frigate will track objects outside of the sidewalk zone, but only the `person` objects appearing in the `sidewalk` zone will show as alerts in the UI.

When creating zones for your cameras, remember that an object's presence in a zone is evaluated based on the bottom center of the bounding box. The [example in the documentation](https://docs.frigate.video/configuration/zones) shows that more clearly.

### Applicable situations for motion masks

Motion masks are used to prevent unwanted types of motion from triggering object detection. A great use for a motion mask would be to cover a camera timestamp, to cover the sky and any moving clouds, or to cover the top of a tree for moving branches and leaves.

### Official Frigate documentation

[Masks](https://docs.frigate.video/configuration/masks)

[Zones](https://docs.frigate.video/configuration/zones)

[Review](https://docs.frigate.video/configuration/review) - for configuring alerts and detections (Frigate 0.14 and later)

---

If this has been a helpful post, give it an upvote and thumbs up. If you have any questions, leave them down below for us! 

-The Frigate dev team (Blake @blakeblackshear, Nick @NickM-27, Josh @hawkeye217)

**Comment by highpeakrider** [comment]
Sorry I'm only just getting my system setup, could you expand on the 'Limit detections to specific zones' element?

**Reply by hawkeye217** [reply]
Check out the official docs here: https://docs.frigate.video/configuration/zones#restricting-alerts-and-detections-to-specific-zones

**Reply by highpeakrider** [reply]
Yes I've read that multiple times but can't see how it covers this function, being new practical examples seem easier to understand rater than code snippets which is why the tip post works well.....

**Reply by hawkeye217** [reply]
What is it specifically that you don't understand? We can try to improve the docs if you think they are not clear.

**Reply by highpeakrider** [reply]
Hi Josh
That's a difficult question, as a complete newbie , non developer who can follow code by example I struggled to understand how to fit it all together and where and how to enter the code. I did this on my Synology NAS but have a dedicated machine that I'm going to try and build using Debian today. On the internet I found lots of You Tube creators but much of the content relates to old versions now which can take you down dead ends.

MGTT was real fun getting it into HA, creating the docker but understanding the password creation was a stumble but I found this to create a password file. https://dmelo.eu/blog/mosquitto_passwd_gen/

I find something like the below really easy to follow, however I'm mindful this is a free opensource app and resources may be limited, and writing documentation was not my favourite thing when I was working, although I'd love to get involved if I understand the subject matter to document it.
easy example - https://mariushosting.com/how-to-install-home-assistant-on-your-synology-nas/

I was struggling with this as it was picking up the road traffic and next doors car on the drive on the right but I've now reduced the area to see if that works better.
Feel free to delete this ramble if it distracts from the tip above.

Cheers Phil
<img width="1145" alt="Screenshot 2024-11-27 at 14 10 47" src="https://github.com/user-attachments/assets/02bf11db-0c42-4534-9916-f2846e254db7">

**Reply by hawkeye217** [reply]
What you've done to change the `drive_zone` is exactly right. Because presence in a zone is [based on the bottom center of an object's bounding box](https://docs.frigate.video/configuration/zones/), cars driving by on the road were likely entering the zone. You can watch the Debug view and enable bounding boxes and zones there to see what happens when cars drive by now.

**Reply by NickM-27** [reply]
Make sure the zone is set as required in the camera settings page

**Reply by highpeakrider** [reply]
Here?
<img width="541" alt="Screenshot 2024-11-27 at 14 34 39" src="https://github.com/user-attachments/assets/333b7edf-19d3-4624-9a60-982ce20221e9">

**Reply by NickM-27** [reply]
have you limited alerts to the zone?

**Comment by Aquathing** [comment]
What about objects that are in between a motion mask? 
I am using motion masks to exclude motion from my hedges, as they trigger object detection which drastically increases CPU usage (I do not have and cannot get a Coral yet). 

If somebody is between the hedges (i.e only the feet are visible outside the motion mask from debug view), would that trigger object detection on the full area of the moving object or only the feet?

**Reply by highpeakrider** [reply]
Where are you based loads in the UK. https://www.mouser.co.uk/ProductDetail/Coral/G950-06809-01?qs=u16ybLDytRbcxxqFKdbhgQ%3D%3D

**Reply by Aquathing** [reply]
I'm in Italy. Yes, I found that website, but I cannot buy a Google Coral right now, I'm trying to decrease CPU usage as to consume fewer watts while maintaining security.

**Reply by NickM-27** [reply]
Once an object is detected motion masks won't affect detection as tracking information is used to know where to look

**Reply by Aquathing** [reply]
> Once an object is detected motion masks won't affect detection as tracking information is used to know where to look

That's great to hear! Thanks.

**Comment by neuviemeporte** [comment]
I think the big red text warning that big motion masks are "not recommended" in the UI should be changed. I am aware of the difference between masks and zones, and I have a required zone on my backyard camera. But I also use a big motion mask to block off the sky and the brush above the lawn because it keeps moving and increases the load on my hardware. No objects that I'm interested in are going to fly into the sky and come back down. After an upgrade, the warning text gaslighted me into thinking there could be a performance penalty from using a big motion mask, but from what I understand the worst it could do is make Frigate generate more detection events? It's good to have tips, but IMO it should read something like "large motion masks could interfere with tracking objects which step into the motion mask", and potentially it should be hidden altogether if there is a required zone also configured. Otherwise, I think it's increasing the confusion rather than alleviating it.

**Reply by NickM-27** [reply]
Many users are not aware of this, and a warning is how they find out about it. Using large masks can affect performance in that object detection is fun more frequently along with worse tracking and potentially missing important things altogether 

> but from what I understand the worst it could do is make Frigate generate more detection events?

No, the worst it can do is cut someone's recording short leaving out details of where a person came from before and where they went after breaking in to the property. It is very much a problem and should be avoided. 

The warning only shows when a massive area of the camera is covered with a mask which in most cases is wrong. In cases like you describe it does make more sense, but often it would seem like adjusting the camera angle may be better to avoid so much resolution pointed at areas that aren't interesting (as a general example, there can of course be reasons it's necessary).

**Comment by rinaldodev** [comment]
**tl;dr:** After using Frigate for 2 months, I feel like the need to define "areas of interest for detection initiation" is very common and legitimate. For users, motion masks feel like they would be the perfect tool for this because we don't realize it loses track of objects that step inside of it.

> Because of overmasking, the person of interest would actually be tracked by Frigate as an entirely new person as they leave the masked area and proceed down the sidewalk the second time, potentially creating two separate review items. In a high-traffic area, this would certainly not be ideal.

I understand why this happens today, but I don't understand what limitations exist that wouldn't allow overcoming this.
Is there a technical limitation in keeping track of all objects that started or passed through an unmasked area?

I'm asking because this seem to be a source of confusion for a lot of users, and usually it comes from the need of masking areas that usually shouldn't initiate detection (like small moving shadows).

In my particular case, I have two cameras outside that capture a lot of motion from moving leafs. Masking the tree and branches is easy, but masking the floor where the leafs' shadows are moving, not so much. Since I don't expect any motion to be initiate exactly inside the leafs' shadows area, only around it, masking it feels like it should be the right solution.

Even though zones can help with this, motion detection feels superior because:

- UI-view of the history is impacted by motion detection
- Recording/storage is impacted by motion detection
- GPU usage (your electric bill) is impacted by motion detection

**Reply by NickM-27** [reply]
There are very few cases where an object truly won't ever originate in an area like that. This is because shadows move with the sun so those areas that would be masked become quite large, objects can be lost while being tracked which then causes the need to re-detect. 

There's also another problem, stationary objects. An object might not originate in an area like you describe but it can certainly become stationary there, and when it starts to move again that motion would be masked out. Stationary objects are not checked every frame for efficiency, so the motion is what helps Frigate know when that object is starting to become active again. 

There's nothing in Frigates implementation that stops you from masking that area, the limitation is simply real world scenarios and edge cases which don't fit into the ideal situations that users assume are the only scenarios to account for.

**Reply by rinaldodev** [reply]
Thanks for the insights, Nick.

> This is because shadows move with the sun so those areas that would be masked become quite large, 

Correct me if I'm wrong, but that feels like a different limitation: that motion masks cannot be dinamically adjusted (apart from doing it manually in the UI)? It could in theory be solved by profiles or some other form of dynamic setting (changing via a MQTT message for example)?

> objects can be lost while being tracked which then causes the need to re-detect.

Again correct me if I'm wrong, but this feels like an issue that isn't tied just to this particular use-case? Since it already happens today with or without motion masks, which sometimes causes things like duplicated detections, etc. At some point the object would move out of the masked zone and get re-detected (it's what already happens today IIUC).

> Stationary objects are not checked every frame for efficiency, so the motion is what helps Frigate know when that object is starting to become active again.

Just thinking out loud, without knowledge of all of Frigate's internals. What I was thinking when writting that message was some solution like: that particular object that got detected not being restricted anymore by any motion masks after detection initiated. But that's coming from a user's perspective.

> the limitation is simply real world scenarios and edge cases which don't fit into the ideal situations that users assume are the only scenarios to account for.

Sorry if that's what it sounded like. I'm not arguing on the way Frigate handles motion masks, it's just the tool that we have today. I'm just trying to raise the general need, not a specific solution. My key point is that wanting a way to ignore motion in a specific area, while *not* influencing detections that originated outside that area, feels like a common and legitimate use case. If or how that can be achieved, without hurting other scenarios, is what I'm trying to understand.

**Reply by NickM-27** [reply]
Profiles are already supported in 0.18, but this would get very complicated considering that times of the year the shadows are different. All the effort really doesn't seem worth it vs just having some extra motion at times. 

> Since it already happens today with or without motion masks, which sometimes causes things like duplicated detections, etc. At some point the object would move out of the masked zone and get re-detected (it's what already happens today IIUC).

Yes, to an extent, but that breaks the original assumption that an object originates somewhere else. Now we are saying that no motion or object record would exist in this case while it was in this motion masked area. An unauthorized person could do anything in this area and it wouldn't be recorded by anything other than continuous recording, if that's enabled. It's the exact real life use case that seems "fine" but in practice has many holes that don't exist when you just let some motion happen. 

> My key point is that wanting a way to ignore motion in a specific area, while not influencing detections that originated outside that area, feels like a common and legitimate use case.

Right, that's technically how it already works today. An object will continue to be tracked if it started outside a **motion** mask and moves through one. But there are so many edge cases, it doesn't make sense to try and work through and account for all these vs just accepting a bit of motion detection which is what Frigate is designed to do efficiently anyway.

So the recommendation is not to motion mask large areas where objects you care about will exist. Then you never have to worry about missing something.

**Reply by rinaldodev** [reply]
> it doesn't make sense to try and work through and account for all these vs just accepting a bit of motion detection which is what Frigate is designed to do efficiently anyway.
> 
> So the recommendation is not to motion mask large areas where objects you care about will exist. Then you never have to worry about missing something.

Thanks, Nick. I think now I understand the philosophy behind it.

I guess the hard pill to swallow for me is that this philosophy is a better fit if you have watt-optmized hardware (like Halo, Coral, N150, etc.), so you can have 24/7 AI detection, even turn off motion recording, and Frigate will do it's job all day for just a few bucks a month. OTOH on a desktop with a dedicated GPU, accepting so much more motion starts to hurt a bit more.

**Reply by NickM-27** [reply]
there's a lot that can be done with a GPU to reduce power usage, like setting manual GPU and Memory clock. For my 5060Ti, when I ran that for Frigate, this cut the power usage down from 50W to 20W. Of course that doesn't compare at all to using something like an NPU in an Intel desktop CPU, which is only a couple watts. 

That said there are many reasons to keep motion recording on, like catching non objects such as tree branches falling, fireworks, fences blowing over, whatever the case may be. Which is another piece of the overall reason for the approach.
